src/algoliasearch.helper.js

```diff
@@ -1223,15 +1223,16 @@ AlgoliaSearchHelper.prototype._dispatchAlgoliaResponse = function(states, queryI
   }
 
   this._currentNbQueries -= (queryId - this._lastQueryIdReceived);
-  if (this._currentNbQueries === 0) this.emit('searchQueueEmpty');
-
   this._lastQueryIdReceived = queryId;
 
   if (err) {
     this.emit('error', err);
+    if (this._currentNbQueries === 0) this.emit('searchQueueEmpty');
     return;
   }
 
+  if (this._currentNbQueries === 0) this.emit('searchQueueEmpty');
```

Comment:
Yes, actually this line is at a lot of places in the code. It could be refactored.
In this function, I went the easy path, and didn't want to change too much to ease the discussion by limiting the diff.

If you have a better way to approach this, please feel free to adapt.

If you think my approach could lead to problems, let's chat about it sometime.

There is no big hurry.